File: 2.add-two-numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    	ListNode ans = new ListNode(0);
    	ListNode p1 = l1;
    	ListNode p2 = l2;
    	ListNode pAns = ans;
    	int carry = 0;
    	while (true)
    	{
    		if (p1 != null)
    		{
    			pAns.val += p1.val;
    			p1 = p1.next;
    		}
    		if (p2 != null)
    		{
    			pAns.val += p2.val;
    			p2 = p2.next;
    		}
			pAns.val += carry;
			carry = pAns.val / 10;
			pAns.val %= 10;
    		if (p1 == null && p2 == null && carry == 0)
    			break;
    		pAns.next = new ListNode(0);
    		pAns = pAns.next;
    	}
    	return ans;
    }
}

File: 2.add-two-numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return addTwoNumbers(l1, l2, 0);
    }

private:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2, int carry) {
        if (l1) {
            if (l2) {
                int val = l1->val + l2->val + carry;
                return new ListNode(val % 10, addTwoNumbers(l1->next, l2->next, val / 10));
            } else {
                int val = l1->val + carry;
                return new ListNode(val % 10, addTwoNumbers(l1->next, nullptr, val / 10));
            }
        } else {
            if (l2) {
                int val = l2->val + carry;
                return new ListNode(val % 10, addTwoNumbers(nullptr, l2->next, val / 10));
            } else {
                return carry > 0 ? new ListNode(carry) : nullptr;
            }
        }
    }
};
